src/lib.rs

```diff
@@ -224,24 +224,31 @@ pub trait Interrupts {
     fn get_interrupts(&mut self, clear: bool) -> Result<Self::Irq, Self::Error>;
 }
 
+/// Register contains the address and value of a register.
+///
+/// It is primarily intended as a type constraint for the [Registers] trait.
+pub trait Register: Copy + From<u8> + Into<u8> {
+    const ADDRESS: u8;
+}
+
 /// Registers trait provides register level access to the radio device.
 ///
 /// This is generally too low level for use by higher abstractions, however,
 /// is provided for completeness.
-pub trait Registers<R: Copy> {
+pub trait Registers {
     type Error: Debug;
 
     /// Read a register value
-    fn reg_read(&mut self, reg: R) -> Result<u8, Self::Error>;
+    fn read_register<R: Register>(&mut self) -> Result<R, Self::Error>;
 
     /// Write a register value
-    fn reg_write(&mut self, reg: R, value: u8) -> Result<(), Self::Error>;
+    fn write_register<R: Register>(&mut self, value: R) -> Result<(), Self::Error>;
 
     /// Update a register value
-    fn reg_update(&mut self, reg: R, mask: u8, value: u8) -> Result<u8, Self::Error> {
-        let existing = self.reg_read(reg)?;
-        let updated = (existing & !mask) | (value & mask);
-        self.reg_write(reg, updated)?;
+    fn update_register<R: Register>(&mut self, f: fn(R) -> R) -> Result<R, Self::Error> {
```

Comment:
the extra generic argument here is a wee bit annoying but i think the closure (`fn reg_update<R: Reg, F: Fn(&mut R)>(&mut self, f: F)`) approach is better than the function pointer (`fn update_register<R: Register>(&mut self, f: fn(R) -> R)`) option because iirc you can't capture state in a function pointer.

i generally prefer `noun_verb` (eg. `reg_write`) over `verb_noun` (eg. `write_reg`) because it means like operations are grouped alphabetically (a stylistic hangover from all the C i write), but i think most of the other traits are `verb_noun` because `check_transmit` was better than `transmit_check` so, the change makes sense.

Reply:
Good point, made it a closure (but taking R by value instead of mutable reference)

Reply:
cool cool, do you have a reason for preferring to take R by value rather than reference? i guess it means you use the `.with` methods which are chainable rather than the `.set` ones, will try swapping my WIP to this ^_^

Reply:
R is going to be u8/u16/u32 approx 99.9% of the time and I wouldn't pass any of those by reference.

If you are lucky, the compiler is able to optimize away the reference, if not, then yeah...

(pass-by-value is also more ergonomic since you have ownership of R in the closure)

Reply:
ahh makes sense, and works just the same for one-liners using `with` instead of `set`